In [3]:
import numpy as np 
import pandas as pd 
from Stacking import CustomStackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split 
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import StackingClassifier
import warnings 

warnings.filterwarnings('ignore')

In [4]:
X, y = make_classification(
    n_samples=10000, 
    n_features=3, 
    n_informative=3,
    n_redundant=0,    
    n_classes=3, 
    random_state=42
)

X = pd.DataFrame(X, columns = [f"Feature {i}" for i in [1,2,3]])

In [5]:
X

,Feature 1,Feature 2,Feature 3
0,-1.652679,-1.656913,1.067450
1,1.107570,-1.112789,0.836552
2,3.380797,1.188618,-4.164533
3,-2.108716,-1.106227,-0.136760
4,-1.900634,-2.090601,1.248259
...,...,...,...
9995,0.800127,-2.785169,-0.759850
9996,1.015938,0.565011,-0.723957
9997,1.743000,1.230476,0.305473
9998,-1.492344,-0.501008,-0.372274


In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

scalar = StandardScaler()

X_train = scalar.fit_transform(X_train)
X_test = scalar.transform(X_test)

In [7]:
estimators = [LogisticRegression(), KNeighborsClassifier(n_neighbors=7), DecisionTreeClassifier()]

stack = CustomStackingClassifier(estimators=estimators, meta_estimator=LogisticRegression())
stack.fit(X_train, y_train)

In [8]:
custom_pred = stack.predict(X_test)

In [9]:
acc = accuracy_score(y_test, custom_pred)
print("accuracy : ", acc)

accuracy :  0.9275757575757576


In [12]:
from sklearn.metrics import f1_score

f1 = f1_score(y_test, custom_pred, average="macro")
print("f1 : ", f1)

f1 :  0.9277035133314943


In [13]:
estimators = [("LinearRgression" , LogisticRegression()), ("KNN",KNeighborsClassifier(n_neighbors=7)), ("DT", DecisionTreeClassifier())]

In [14]:
sk_stack = StackingClassifier(estimators=estimators, final_estimator=LogisticRegression())
sk_stack.fit(X_train, y_train)


StackingClassifier(estimators=[('LinearRgression', LogisticRegression()),
                               ('KNN', KNeighborsClassifier(n_neighbors=7)),
                               ('DT', DecisionTreeClassifier())],
                   final_estimator=LogisticRegression())

In [16]:
sk_pred = sk_stack.predict(X_test)

sk_acc = accuracy_score(y_test, sk_pred)

print("Accuracy : ", sk_acc)

sk_f1 = f1_score(y_test, sk_pred, average="macro")

print("sk_rmse : ", sk_f1)

Accuracy :  0.93
sk_rmse :  0.9301327786331659


In [17]:
pd.DataFrame({
    'custom' : [acc,f1],
    'sklearn': [sk_acc,sk_f1]
}, index=['Accuracy', 'F1_score'])

,custom,sklearn
Accuracy,0.927576,0.930000
F1_score,0.927704,0.930133


## Conclusion

I implemented a Stacking Classifier from scratch using multiple base classifiers and a Logistic Regression meta-model. I compared it with sklearn's `StackingClassifier` using the same setup. My custom model achieved an accuracy of **0.927576** and an F1-score of **0.927704**, while sklearn achieved an accuracy of **0.930000** and an F1-score of **0.930133**. Both models produced almost identical results, showing that my scratch implementation correctly captures the main working of Stacking Classification.

